### ISSUE TITLE
breaking plugin API change: Formalize strict extensibility mechanism

### ISSUE THREAD

**Issue by atsushieno** [atsushieno/android-audio-plugin-framework#103]
While I'm revisiting presets extensibility (issue #51), I noticed that current AAP documentation (`README.md`) has poor wording on extensibility, and moreover, it does not really provide the whole mechanism and real-world usage -

- aap::SharedMemoryExtension used to be regarded as one of them, but it is not for public anymore (and it should not be). 
- MIDI2 extension exists, but it is not going to be useful at this moment (as we could simply implement MIDI-CI based protocol promotion without any extensibility), and feature wise it does not require any function either.

Presets are different. We certainly need some functions implemented by plugin (if we support presets metadata then we can do it without functions, but that is not intuitive for plugin developers as `get_state()` and `set_state()` are implemented as C functions).

Then I noticed that there is no `get_extension(URI)` for the plugin (`extension_data()` in LV2). I must have thought it is whatever `instantiate()` in `AndroidAudioPluginFactory` can pass (which is wrong; its `extensions` argument is a list of **host** extensions, not plugin extensions). To add this, we have to break the plugin API for the first time in these two years (second in three years).

And going back to README, we should clarify that extensibility should not prohibit functions. They can be defined, just need to make it clear that they are not invoked beyond applications border.

We should also clarify whether the extensions data (which consists of shared pointers) is mutable or not, shared across plugin instances, and/or across services.

In any case this is a big breaking change that invalidates any of the existing clients and hosts, it is probably a big opportunity to explore how to introduce breaking plugin API changes.

**Comment by atsushieno** [comment]
It will be included in 0.7.2 once `breaking-extensibility-change` branch is merged.

In addition to the changes described above, there will be a new plugin API type `AndroidAudioPluginHost` and it is passed to `instantiate()` function pointer in the plugin API, instead of `AndroidAudioPluginExtension**`.

Also, `AndroidAudioPluginFactory` now has a `factory_context` field that can be used by `insstantiate()`.

**Comment by atsushieno** [comment]
The changes are now merged into main.
